### PR TITLE
Known Hosts

### PR DESCRIPTION
**Pull request by ejohnstown**
Added Known Hosts file checking for the wolfssh application. It currently checks the file "~/.ssh/known_hosts". Reports missing, duplicate, or bad keys.

### CODE REVIEW

File: apps/wolfssh/common.c [hunk 1]
```diff
@@ -186,20 +187,179 @@ void ClientIPOverride(int flag)
 #endif /* WOLFSSH_CERTS */
 
 
+static int AppendKeyToFile(const char* filename, const char* name,
+        const char* type, const char* key)
+{
+    FILE *f;
+
+    f = fopen(filename, "a");
```
Comment (JacobBarthelmeh) [inline apps/wolfssh/common.c:195]: needs to be WFOPEN (remapped for Windows builds)

File: apps/wolfssh/common.c [hunk 2]
```diff
@@ -186,20 +187,179 @@ void ClientIPOverride(int flag)
...
+
+
+static int FingerprintKey(const byte* pubKey, word32 pubKeySz, char* out)
+{
+    wc_Sha256 sha;
+    byte digest[WC_SHA256_DIGEST_SIZE];
+    char fp[48] = { 0 };
+    word32 fpSz = sizeof(fp);
+
+    wc_InitSha256(&sha);
```
Comment (JacobBarthelmeh) [inline apps/wolfssh/common.c:210]: checks on return values

File: apps/wolfssh/common.c [hunk 3]
```diff
@@ -186,20 +187,179 @@ void ClientIPOverride(int flag)
...
+    char *line;
+    const char *targetName = (const char*)ctx;
+    char *name;
+    char *keyType;
+    char *key;
+    char *knownHosts = NULL;
+    char *knownHostsName = NULL;
+    int ret = 0, found = 0, badMatch = 0, otherMatch = 0;
+    word32 sz, lineCount = 0;
+    char encodedKey[1200] = { 0 };
```
Comment (JacobBarthelmeh) [inline apps/wolfssh/common.c:254]: pretty large stack value and magic number, needs a small stack option and macro or comment on value choosen

File: apps/wolfssh/common.c [hunk 4]
```diff
@@ -186,20 +187,179 @@ void ClientIPOverride(int flag)
...
-    #else
-        (void)pubKey;
-        (void)pubKeySz;
-        (void)ctx;
-    #endif
+    sz = 0;
+    ret = load_der_file(knownHostsName, (byte**)&knownHosts, &sz);
+
+    /* Get the key type out of the key. */
+    ato32(pubKey, &sz);
```
Comment (JacobBarthelmeh) [inline apps/wolfssh/common.c:275]: sanity check on 'sz' for size of pubKey buffer before use

File: apps/wolfssh/common.c [hunk 5]
```diff
@@ -186,20 +187,179 @@ void ClientIPOverride(int flag)
...
+    pubKeyType[sz] = 0;
+
+    sz = (word32)sizeof(encodedKey);
+    ret = Base64_Encode_NoNl(pubKey, pubKeySz, (byte*)encodedKey, &sz);
+    FingerprintKey(pubKey, pubKeySz, fp);
+
+    cursor = knownHosts;
+    while (cursor) {
+        lineCount++;
+        line = WSTRSEP(&cursor, "\n");
```
Comment (JacobBarthelmeh) [inline apps/wolfssh/common.c:286]: Heap buffer overflow can happen around this area of code, please add additional sanity checks on upper bounds of cursor.

**Review by JacobBarthelmeh — CHANGES_REQUESTED** [review]

File: apps/wolfssh/common.c (hunk 3 above)
Comment (ejohnstown) [reply apps/wolfssh/common.c:254]: The stack wasn't a concern anywhere else in the application. I think this is a case where it should be allocated, though.

**Review by JacobBarthelmeh — APPROVED** [review]

**Pull request merged**